File: Engine/Source/Runtime/Core/IO/FileSystemReference.cpp
// Copyright 2020-2021 Aumoa.lib. All right reserved.

#include "pch.h"
#include "FileSystemReference.h"
#include "DirectoryReference.h"

using namespace std;

FileSystemReference::FileSystemReference(const filesystem::path& filepath) : _path(filesystem::absolute(filepath))
{
}

bool FileSystemReference::IsExists() const
{
	return filesystem::exists(_path.value_or(L"\\"));
}

filesystem::path FileSystemReference::GetPath() const
{
	return _path.value_or(L"\\");
}

DirectoryReference FileSystemReference::GetParent() const
{
	return _path.value_or(L"\\").parent_path();
}

bool FileSystemReference::IsSet() const
{
	return _path.has_value();
}

File: BuildTool/DotNETUtilities/IO/FileSystemReference.cs
// Copyright 2020-2021 Aumoa.lib. All right reserved.

namespace DotNETUtilities;

using System.IO;

public abstract class FileSystemReference
{
    public FileSystemReference(string InPath)
    {
        FullPath = Path.GetFullPath(InPath);
        Name = Path.GetFileNameWithoutExtension(FullPath);
        Root = Path.GetPathRoot(FullPath);
    }

    public string GetRelativePath(FileSystemReference InReference)
    {
        return Path.GetRelativePath(FullPath, InReference.FullPath);
    }

    public DirectoryReference GetParent()
    {
        return new(Directory.GetParent(FullPath).FullName);
    }

    public override string ToString()
    {
        return FullPath;
    }

    public string Name { get; }

    public string FullPath { get; }

    public string Root { get; }
}